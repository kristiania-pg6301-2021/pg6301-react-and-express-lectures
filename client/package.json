{
  "name": "client",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "parcel src/index.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router": "^5.2.0",
    "react-router-dom": "^5.2.0"
  },
  "devDependencies": {
    "parcel": "^2.0.0-beta.2",
    "prettier": "^2.2.1"
  }
}
