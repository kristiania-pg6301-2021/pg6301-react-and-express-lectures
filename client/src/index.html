<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>KristianiaQuiz</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
