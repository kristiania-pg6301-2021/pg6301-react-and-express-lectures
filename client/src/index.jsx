import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>KristianiaQuiz</Link>
      </header>
      <main>
        <Switch>
          <Route path="/login">
            <h1>Login</h1>
          </Route>
          <Route path="/" exact>
            <h1>Home page</h1>
            <ul>
              <li>
                <Link to={"/login"}>Login</Link>
              </li>
            </ul>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
