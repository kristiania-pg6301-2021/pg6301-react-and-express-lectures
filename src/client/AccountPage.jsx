import React, { useEffect, useState } from "react";

function useLoader(loader, deps) {
  const [loading, setLoading] = useState();
  const [data, setData] = useState();
  const [error, setError] = useState();

  async function reload() {
    setLoading(true);
    setError(undefined);
    setData(undefined);
    try {
      setData(await loader());
    } catch (e) {
      setError(e);
    } finally {
      setLoading(false);
    }
  }

  useEffect(reload, deps || []);

  return { loading, data, error, reload };
}

export function AccountPage() {
  const { loading, data, error } = useLoader(async () => {
    const res = await fetch("/api/account");
    if (!res.ok) {
      throw new Error("Failed to GET " + res.url + ": " + res.statusText);
    }
    return res.json();
  });

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>Error: {error.toString()}</div>;
  }

  return (
    <>
      <h1>Account</h1>
      <div>You have ${data.account.balance}</div>
    </>
  );
}
