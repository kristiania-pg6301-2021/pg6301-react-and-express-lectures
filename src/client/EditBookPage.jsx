import React, { useState } from "react";
import { useLocation, useParams } from "react-router";
import { LoadingView } from "./LoadingView";
import { InputField } from "./InputField";
import { useLoader } from "./useLoader";

function EditBookForm({ book, onSubmit }) {
  const [title, setTitle] = useState(book.title);
  const [author, setAuthor] = useState(book.author);
  const [year, setYear] = useState(book.year);
  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        onSubmit({ title, author, year });
      }}
    >
      <h1>Edit book: {title}</h1>
      <InputField label={"Title"} value={title} onChangeValue={setTitle} />
      <InputField label={"Author"} value={author} onChangeValue={setAuthor} />
      <InputField
        label={"Year"}
        value={year}
        onChangeValue={setYear}
        type="number"
      />
      <button>Submit</button>
    </form>
  );
}

export function EditBookPage({ bookApi }) {
  const { search } = useLocation();
  const params = useParams();
  console.log(params);

  const { error, data: book } = useLoader(() => {
    const id = new URLSearchParams(search).get("id");
    return bookApi.fetchBook(id);
  }, [search]);

  async function handleSubmit(book) {
    const id = new URLSearchParams(search).get("id");
    await bookApi.updateBook(id, book);
  }

  if (error) {
    return <div>An error occurred: {error.toString()}</div>;
  }
  if (!book) {
    return <LoadingView />;
  }

  return <EditBookForm book={book} onSubmit={handleSubmit} />;
}
