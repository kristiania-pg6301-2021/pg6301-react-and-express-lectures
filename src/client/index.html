<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>Hello Kristiania</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
