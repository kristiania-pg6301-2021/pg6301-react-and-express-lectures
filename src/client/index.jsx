import * as React from "react";
import * as ReactDOM from "react-dom";
import { Application } from "./Application";

ReactDOM.render(<Application />, document.getElementById("app"));
