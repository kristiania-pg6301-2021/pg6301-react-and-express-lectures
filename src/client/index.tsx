import * as React from "react";
import * as ReactDOM from "react-dom";

import {BrowserRouter} from "react-router-dom";
import {Route, Switch} from "react-router";

interface Quiz {
    question: string;
    alternatives: string[];
}


function HomePage() {
    return <>
        <h1>Welcome to the quiz</h1>
        <button>Start new quiz</button>
    </>
}

function QuizPage({quizFactory}: {quizFactory(): Quiz}) {
    const quiz = quizFactory();
    const {question, alternatives} = quiz;
    return <>
        <h1>Quiz</h1>
        <div>Question: {question}</div>
        {alternatives.map(a =>
            <button>{a}</button>
        )}
    </>
}

function App() {
    const quiz: Quiz = {
        question: "Are you happy?",
        alternatives: ["yes", "no", "maybe"]
    }
    return <BrowserRouter>
        <Switch>
            <Route path="/quiz">
                <QuizPage quizFactory={() => quiz} />
            </Route>
            <Route exact path="/">
                <HomePage />
            </Route>
            <Route>
                Not found
            </Route>
        </Switch>
    </BrowserRouter>;
}


ReactDOM.render(<App />, document.getElementById("app"));