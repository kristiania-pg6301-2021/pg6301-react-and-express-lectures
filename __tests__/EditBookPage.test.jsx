import { EditBookPage } from "../src/client/EditBookPage";
import { act, Simulate } from "react-dom/test-utils";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { MemoryRouter, Route } from "react-router";

async function render(component, url = "/books/12/edit") {
  const container = document.createElement("div");
  await act(async () => {
    await ReactDOM.render(
      <MemoryRouter initialEntries={[url]}>
        <Route path={"/books/:id/edit"}>{component}</Route>
      </MemoryRouter>,
      container
    );
  });
  return container;
}

describe("book edit page", function () {
  it("loads book", async () => {
    const getBook = () => ({
      title: "My Book",
      author: "Firstname Lastname",
      year: 1999,
    });
    const container = await render(<EditBookPage bookApi={{ getBook }} />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("shows book loading failure", async () => {
    const getBook = async () => {
      throw new Error("Not found");
    };
    const container = await render(<EditBookPage bookApi={{ getBook }} />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it("updates book on save", async () => {
    const getBook = () => ({
      title: "My Book",
      author: "Firstname Lastname",
      year: 1999,
    });
    const updateBook = jest.fn();
    const container = await render(
      <EditBookPage bookApi={{ getBook, updateBook }} />
    );

    Simulate.change(container.querySelector("input"), {
      target: {
        value: "New Value",
      },
    });
    Simulate.submit(container.querySelector("form"));
    expect(updateBook).toHaveBeenCalledWith("12", {
      title: "New Value",
      author: "Firstname Lastname",
      year: 1999,
    });
  });
});
