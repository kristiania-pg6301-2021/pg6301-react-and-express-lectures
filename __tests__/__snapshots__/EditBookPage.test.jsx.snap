// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book edit page loads book 1`] = `"<form><h1>Edit an existing book (My Book)</h1><div><label>Title: <input type=\\"text\\" value=\\"My Book\\"></label></div><div><label>Author: <input type=\\"text\\" value=\\"Firstname Lastname\\"></label></div><div><label>Year: <input type=\\"number\\" value=\\"1999\\"></label></div><button>Submit</button></form>"`;

exports[`book edit page shows book loading failure 1`] = `"<div>Something went wrong: Error: Not found</div><button>Try again</button>"`;
